src/Winton.Extensions.Configuration.Consul/ConsulConfigurationProvider.cs

```diff
@@ -27,30 +30,37 @@ internal sealed class ConsulConfigurationProvider : ConfigurationProvider
 
             _consulConfigClient = consulConfigClient;
             _source = source;
+            _cancellationTokenSource = new CancellationTokenSource();
 
             if (source.ReloadOnChange)
             {
-                ChangeToken.OnChange(
-                    () => _consulConfigClient.Watch(_source.Key, _source.OnWatchException, _source.CancellationToken),
+                _changeTokenRegistration = ChangeToken.OnChange(
+                    () => _consulConfigClient.Watch(_source.Key, _source.OnWatchException, _cancellationTokenSource.Token),
                     async () =>
                     {
-                        await DoLoad(true).ConfigureAwait(false);
+                        await DoLoad(true, _cancellationTokenSource.Token).ConfigureAwait(false);
                         OnReload();
                     });
             }
         }
 
         public override void Load()
         {
-            DoLoad(false).GetAwaiter().GetResult();
+            DoLoad(false, _cancellationTokenSource.Token).GetAwaiter().GetResult();
         }
 
-        private async Task DoLoad(bool reloading)
+        public void Dispose()
+        {
+            _cancellationTokenSource?.Cancel();
+            _changeTokenRegistration?.Dispose();
+        }
+
+        private async Task DoLoad(bool reloading, CancellationToken token)
         {
             try
             {
                 QueryResult<KVPair[]> result = await _consulConfigClient
-                    .GetConfig(_source.Key, _source.CancellationToken)
+                    .GetConfig(_source.Key, token)
```

Comment:
Given this is an instance method you could access `_cancellationTokenSource.Token` here and simplify the signature to this method for the other callers in the class.

Reply:
It could. However I've done this because I'm anticipating a further refactor away from using the change token registration. Hopefully I'll get a chance to finish this PR off soon.